issue 21 -- Configuration Registry initial implementation

NorthWindH opened the pull request:
Initial implementation of configuration registry ( aside from load/store ) is ready for peer review. I tried to follow style guidelines exhibited in other files as much as possible.

files of note:
`src/gladius-src.js`
`src/core/Configurator.js`
`test/core/Configurator.js`

Please note the huge hack in src/core/Configurator.js:148 and the associated hack in test/core/Configurator.js:118, ideas on a better way to test construction/defaultConfig loading?

Issue thread: https://github.com/alankligman/gladius/issues/21

NorthWindH committed:
Added some configurator tests, most prominently get/set

NorthWindH committed:
Added getPath tests

NorthWindH committed:
Added configurator tests to index.html

NorthWindH committed:
Added default configuration test

NorthWindH committed:
Minor updates.

NorthWindH committed:
Partial listen/ignore test.

NorthWindH committed:
Updated some status messages

NorthWindH committed:
Reordered tests

NorthWindH committed:
Added listen/ignore tests

NorthWindH committed:
Added set JSON test

NorthWindH committed:
Moved a few expect statements around, making code prettier

NorthWindH committed:
Added header blurb to test suite.

NorthWindH committed:
Beginning work on configurator development.

NorthWindH committed:
Updated set( JSON ) test

NorthWindH committed:
Added get function and partial definition of node

NorthWindH committed:
Refactored traversal function

NorthWindH committed:
Added set function

NorthWindH committed:
Updated default configuration test to acquire configurator better

NorthWindH committed:
Updated configurator construction not to be partially constructed
Updated traverse to enable path creation
HACK constructor reference stored on every object...

NorthWindH committed:
Added Configurator to Gladius' construction sequence

NorthWindH committed:
Reorganized src/Configurator.js

NorthWindH committed:
Implemented getPath

NorthWindH committed:
Implemented listen/ignore

NorthWindH committed:
Added another test to listen/ignore

NorthWindH committed:
Updated some minor quibbles to pass check-lint

humphd commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,225 @@
+        
+        var self = this;
+        
+        self.name = name;
```
Use this vs. self for attaching instance properties.  Stashing this on self is a good idea, but don't use it everywhere.

humphd commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,225 @@
+            
+            var rv;
+            
+            if ( path.length == 1 && path.charAt( 0 ) == '/' ) {
```
=== vs. ==

humphd commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,225 @@
+                // Parse path and traverse the node tree
+                var pathElems = path.split('/');
+                var curNode = self;
+                var successful = true;
```
var a,
  b,
  c;

vs. using var var var

humphd commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,225 @@
+                var pathElems = path.split('/');
+                var curNode = self;
+                var successful = true;
+                for ( var i = 0; successful && i < pathElems.length; ++ i ) {
```
cache pathElems.length:

for (var i=0, iMax = pathElems.length; i < iMax; ++i)

humphd commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,225 @@
+            
+            if ( self.parent ) {
+                // Clean last slash if present
+                if ( path.length == 1 && path.charAt( 0 ) == '/' ) {
```
=== vs. ==

humphd commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,225 @@
+                path = '/' + childName + path;
+                
+                // Call all of our listeners
+                for ( var key in self.listeners ) {
```
when possible, avoid for..in loops, as they are more expensive.

NorthWindH replied:
In this case, how do we perform the same operation? `self.listeners` is an associative array keyed with string ids. Should I be using `Object.keys()`? Is that not just as expensive? Is there a quicker alternative?

humphd commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,225 @@
+                
+                // Call all of our listeners
+                for ( var key in self.listeners ) {
+                    // is hasOwnProperty desired here? Seems unnecessary
```
I would, since you don't want to work with keys in the prototype.

humphd commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,225 @@
+     * Borrowed from Sagi Shkedy's Technical Blog
+     * http://blog.shkedy.com/2007/01/createing-guids-with-client-side.html
+     */
+    var _incCounter = 0;
```
Should code for getting a guid not live somewhere more central in gladius?

NorthWindH replied:
I'm inclined to think that it should. From what I could tell ( I will preemptively declare that I did not put much time into looking :) ), such code does not already exist in an accessible location.

modeswitch replied:
There's already code for doing this. Use engine.nextGUID. You're not serializing it, so it should be safe.

NorthWindH replied:
sounds good, I will update the configurator to be partially constructing in order to allow this.

edit: [done](https://github.com/northWind87/gladius/blob/develop_issue_21/src/core/Configurator.js#L145)

humphd commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,225 @@
+        var self = this;
+        
+        var init = function() {
+            var self = this;
```
not necessary, since self exists in the closure around init.

NorthWindH replied:
Are you sure? In [src/core/Configurator.js:184](https://github.com/northWind87/gladius/blob/develop_issue_21/src/core/Configurator.js#L184), I use `new init()` to use init as a constructor of sorts. In this case, I want `self` to refer to the new object, not the object onwhich `getPath()` is running.

edit: `init()` removed

humphd commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,225 @@
+            // Update configuration with given json object
+            self.update = function( json ) {
+                for ( var key in json ) {
+                    if (json.hasOwnProperty( key )) {   // Performance Note: perhaps protecting against the prototype is not required?
```
leave this if doing for...in

humphd commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,225 @@
+            // If provided, associate listenerFunc with the newly created configurator client.
+            self.getPath = function( path, listenerFunc ) {
+                var targetNode = self.node.traverse( path, true ),
+                    rv = new init();
```
init is a ctor?  That's a bad name then.  ctor functions shoudl start with a capital.

humphd commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,225 @@
+            };
+        };
+        
+        init.apply(self);
```
apply is slower than using a direct function call, can you get around this?

humphd commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,225 @@
+        
+        init.apply(self);
+        
+        self.node = new ConfNode( 'ROOT', undefined );  // Associate client front end with tree root
```
don't pass undefined, just leave the arg out: new ConfNode( 'ROOT' );

NorthWindH committed:
Replaced use of self with this in most cases

NorthWindH committed:
Replaced use of == with === in most cases

NorthWindH committed:
Removed use of repeated vars

NorthWindH committed:
Now relevant for loops cache <array>.length value before looping

NorthWindH committed:
Added hasOwnProperty safeties to for..in loops

NorthWindH committed:
Removed some unnecessary comments

NorthWindH committed:
Reorganized Configurator constructor function, removed init

NorthWindH commented:
Resolved all comments.

[Diff of changes](https://github.com/northWind87/gladius/compare/02688016...82292e15909e792812a312032d1788cfd8d9e8a6)

modeswitch commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,218 @@
+        
+        this.node = node || new ConfNode( 'ROOT' );
+        
+        this.constructor = Configurator;    // TODO: FIXME Dirty Dirty Hack
```
Why is this necessary?

NorthWindH replied:
@alankligman
This is actually a problem that I ran into during testing. Essentially, in order to test loading a JSON object during construction, I needed a way to create a configurator object on demand inside the test ([L118](https://github.com/alankligman/gladius/pull/52/files#L2R118)). At the moment, the constructor is not stored in a good place. Would it be a good idea to store a reference to the configurator constructor somewhere on engine? Perhaps the test should be changed? What do you think.

modeswitch replied:
You can use `Object.getPrototypeOf( engine.configurator ).constructor` to get Configurator, which you can then use to make a new instance by `var myConfigurator = new Configurator( engine )`.

NorthWindH replied:
Nice, that sounds perfect. I'll complain if I blow something up.

edit: Resolved. Updated construction method, now partially constructed.

NorthWindH committed:
Removed unnecessary constructor stomping

NorthWindH commented:
Ready for more review. Also looks like I've angered the end of line gods... my apologies.

edit: alright cleaned up, new diff

[diff of changes](https://github.com/northWind87/gladius/compare/82292e15909e792812a312032d1788cfd8d9e8a6...aa755ccd1215f76c13518cf6d651a3b14a542515)

NorthWindH committed:
Moved some funcs around

NorthWindH committed:
Fixed listener bug where listeners were not being
 notified when their bound node had its value changed.

NorthWindH committed:
Updated configurator to be partially constructing.
Configurator now stores root of configuration registry on gladius engine instance.
Added configurator partialctor to engine.core.
Configurator now uses instance nextGUID system.
Fixed some tests where actual/expected values were flipped.

NorthWindH committed:
Added spec standards for getJSON, clear, store, load
Added test for getJSON

NorthWindH committed:
Added clear test

NorthWindH committed:
Added load/store test.

NorthWindH committed:
Added parent element protection tests to clear tests.
Added local storage overwriting tests to load/store tests.

NorthWindH committed:
Added deeper element protection tests to load/store tests.

NorthWindH committed:
Moved load,store,getJSON,clear specs from test to src

NorthWindH committed:
Implemented getJSON

NorthWindH committed:
Implemented store/load

NorthWindH committed:
Fixed up some linting errors

NorthWindH commented:
Implemented load, store, getJSON, clear
At the moment, load only loads from local storage ie a cookie

Review desired. I'll also be posting up some questions pointing to the code shortly.

NorthWindH commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,412 @@
+                // TODO HACK this is extremely dangerous
+                // Do we have any dependencies that handle JSON parsing?
+                // Do browsers reliably provide native JSON parsing?
+                loadedJSON = eval(
```
In the words of check-lint, eval is evil. Do we have better options for parsing JSON?

JSilver99 replied:
Isn't there JSON.parse(str) for that? eval parses any JavaScript, and you probably don't want that.
https://developer.mozilla.org/en/JSON

NorthWindH replied:
@JSilver99
Yeah I'm quite wary of using eval; this is code that can't go into the mainline. With respect to JSON.parse(), how cross-browser compatible is it? I could not find a source that definitively states where JSON is available.

JSilver99 replied:
It looks like any recent stable release of any major browser accepts it.
A quick google brought up this page: http://caniuse.com/json

NorthWindH replied:
Done and done! da80e6a8d47984d92fba6822ec78e5c712f55c62

NorthWindH commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,412 @@
+        defaultConfiguration = defaultConfiguration || {};
+
+        var self = this,
+            _gladiusCookieName = 'gladius_registry',
```
Cookie name for local storage is fixed in stone. Is this OK?

modeswitch replied:
The configurator should accept the cookie name as a parameter during construction.

NorthWindH replied:
On construction, configurator currently accepts a single defaultConfiguration value which is interpreted as a registry JSON object and loaded into the configurator immediately. How about we make the configurator accept a standard options-style object with defaultConfiguration, cookieName and cookieLifetime properties?

modeswitch replied:
That makes sense. Bake a minimal default configuration into the engine that we can pass in to the configurator.

NorthWindH replied:
sounds good.

NorthWindH commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,412 @@
+            // Cookie manipulation functions, slightly modified.
+            // Originally found on Peter-Paul Koch's site, quirksmode
+            // http://www.quirksmode.org/js/cookies.html
+            createCookie = function(name,value,days) {
```
Cookie manipulation functions borrowed from inter-webs. Is having them here OK? Should they be abstracted to another module?

modeswitch replied:
Refactor this into a separate module. You can put the code in src/common.

NorthWindH committed:
Added merge store test

NorthWindH committed:
Fixed getParentPath bug that was adding root node's name

NorthWindH committed:
Fixed load bug that didn't account for case where storage cookie didn't
exist

NorthWindH committed:
Abstracted internal stored json loading.
Updated store to comply with merge store.

NorthWindH committed:
Fixed minor linting error.

NorthWindH committed:
Replaced now using JSON.parse/stringify instead of makeshift stringify and
eval.

NorthWindH committed:
Removed non-relevant comment.

modeswitch commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,416 @@
+     */
+    var ConfNode = function( name, parent ) {
+        
+        var self = this;
```
Use 'that' to store the current context instead of 'self'.

modeswitch commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,416 @@
+
+        // Serializes this node and all of its children as JSON
+        this.getJSON = function() {
+            var rv = {}, children = this.children;
```
Newline after comma.

modeswitch commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,416 @@
+        // Traverse the node tree given a path
+        this.traverse = function( path, doCreatePath ) {
+            
+            var rv;
```
Give this variable a better name (more descriptive).

modeswitch commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,416 @@
+     * Loads and stores configuration data. Allows external code to listen for
+     * changes in configuration subtrees.
+     *
+     * In Gladius, a configurator can be obtained in 3 ways:
```
Remove the rest of this comment below this line.

modeswitch commented on test/core/Configurator.js:
```diff
@@ -0,0 +1,502 @@
+ *
+ * A test suite for the Configurator subsystem.
+ *
+ * Initially written by Hasan (northWind) Kamal-Al-Deen
```
Remove this.

NorthWindH committed:
Changed self references to 'that'

NorthWindH committed:
Replaced var rv with targetNode in node.traverse

NorthWindH committed:
Newline after comma Configurator line 111

NorthWindH committed:
Truncated long comment block.

NorthWindH committed:
Removed vanity line

NorthWindH committed:
Refactored cookie functions into their own module.

NorthWindH committed:
Updated configurator construction parameters, now accepts options object,
see source.

NorthWindH commented:
Implemented all suggestions up until now.

NorthWindH commented on src/common/Cookie.js:
```diff
@@ -0,0 +1,42 @@
+    };
+};
+
+window.gladiusCookie = new _Cookie();
```
window.cookie ( window.math style ) is perhaps vague, too similar to document.cookie
window.gladiusCookie is perhaps not the best name in the world but its not terrible, is a better name desired?

modeswitch replied:
Why is this being stored on the window instance?

NorthWindH replied:
Followed the style of Math.js, where is the preferred storage location.

modeswitch replied:
That's not the correct thing to do here. If I create multiple engines in the same page, this code will break.

NorthWindH commented on test/core/Configurator.js:
```diff
@@ -0,0 +1,500 @@
+        }
+    });
+
+    // Test clear
```
All of the tests from here on-ward have no comment as their role is relatively obvious; I don't like this but I don't hate it enough to write a comment for each of 50+ tests. This can be changed if desired. Please indicate preference.

modeswitch replied:
Each of the tests (equal, ok, etc) needs a string associated with it that makes clear the intent of the test. That string is the only information available on the test page if the tests fail.

NorthWindH replied:
No problem.

NorthWindH commented on src/gladius-src.js:
```diff
@@ -36,7 +37,11 @@ define( function ( require ) {
 
         this.options = options || {};
+        if ( !this.options.defaultConfiguration )   this.options.defaultConfiguration = {};
+        if ( !this.options.cookieName )             this.options.cookieName = 'gladius_registry';
```
Threw cookieName and cookieLifetime onto gladius.options. Is this desirable?

modeswitch replied:
The cookie name can go into the default config. Also, don't assign a default cookie name. If one isn't provided, then we will avoid storing the registry.

modeswitch commented on src/core/Configurator.js:
```diff
@@ -0,0 +1,374 @@
+
+        // Builds a parent path for this node
+        this.getParentPath = function() {
+            rv = '';
```
Please fix these variable names (and similar ones) so that they are clearer.

NorthWindH committed:
Added gitignore.

NorthWindH committed:
Merge remote-tracking branch 'ack/develop' into develop_issue_21

Conflicts:
	.gitignore
	src/gladius-src.js
	src/gladius.js
	test/index.html

NorthWindH committed:
Updated getPath test comments
Fixed bug in getPath test introduced after configurator changed construction style in merge.

NorthWindH committed:
Updated test comments in getJSON

NorthWindH committed:
Clarified getJSON test a bit

NorthWindH committed:
Updated test comments for testClear function

NorthWindH committed:
Hacked window.cookie back in to be removed when cookie architecture is re-worked

NorthWindH committed:
Updated comments on test local load/store func

NorthWindH committed:
Corrected minor linting error

NorthWindH committed:
Updated comments on merge store test

NorthWindH committed:
Renamed vague local variable names in various functions to clearer versions

NorthWindH committed:
Made load/store perform asynchronously

NorthWindH committed:
Added gameID
Added config/default.js
Engine callback is now called when configurator is done loading from local registry

NorthWindH committed:
Now we use IndexedDB instead of cookies. Implementation still buggy, run test suite with python -m SimpleHTTPServer 8000 for more details

NorthWindH committed:
Removed traces of cookie.js

NorthWindH committed:
Load/store and merge tests now account for asynchronicity correctly, previous implementation did not work correctly when true asynchronicity was added( indexeddb )

NorthWindH committed:
Merge remote-tracking branch 'ack/develop' into develop_issue_21

Conflicts:
	src/gladius-src.js
	test/core/engine.js
	test/core/entity.js
	test/index.html

NorthWindH committed:
Renamed Configurator.js to configurator.js in src/core and test/core
Updated references where relevant

NorthWindH commented:
Whew! All suggestions thus far brought up have been implemented and as a result, review requested!

## Details
- All suggested changes upto now have been implemented, see commit messages for details.
- Configurator now uses indexeddb as load/store backend.
- Code has been brought up to date, ack/develop merged into incoming branch.

modeswitch commented on src/gladius-src.js:
```diff
@@ -130,10 +143,12 @@ define( function ( require ) {
-            if (callback) {
-                callback(this);
-            }
+            // Load any persistent configuration then let caller know the engine instance is ready.
```
The config needs to be loaded earlier than this. All services are started by this point. Should be the first thing to do, even before _math is created.

modeswitch commented:
There's a merge conflict in gladius-src, you will need to sync from develop.

NorthWindH committed:
Merge remote-tracking branch 'ack/develop' into develop_issue_21.
Updated gladius startup subroutine to load configurator first and foremost.

Conflicts:
	src/gladius-src.js

NorthWindH commented:
Review requested.
Merge conflict resolved.
Configurator now loads before anything else.

NorthWindH commented:
Configurator dies if db is not allocated. This behaviour should be amended to not use a db instead.

Commit:
Minor fixes.

NorthWindH commented:
Engine no longer provides nextGUID, Configurator should require lang.js and use window.guid

NorthWindH commented:
rootConfNode shouldn't be stored on engine, pass the root configurator in as a constructor parameter instead

NorthWindH commented:
Have load accept named parameter list

Commit:
A fix and some reorganization.

NorthWindH committed:
Updated store/load to accept named parmeters object

NorthWindH committed:
Updated tests so that undefined guid on configurators will fail the suite

NorthWindH committed:
Configurator now uses window.guid() to create a guid, requires lang
Fixed listen/ignore tests, argument to stop is flaky...

NorthWindH committed:
Moved ConfNode out into its own module definition

NorthWindH committed:
minor fix

NorthWindH committed:
Updated configurator not to store configurator constructor on engine.
Updated configurator not to store rootConfNode on engine instance.

NorthWindH committed:
Configurator now disables db usage gracefully if not available

NorthWindH committed:
Minor

NorthWindH commented:
Review requested.
Identified problems resolved.

NorthWindH committed:
Updated db open fail message.

modeswitch commented on test/core/configurator.js:
```diff
@@ -0,0 +1,602 @@
+(function() {
+
+    var engine = null,
+        getNoDBMsg = function() {
```
Why is this a function? Also, there's already a mechanism to start up a server to run the tests (make test) so just instruct people to use that.

NorthWindH replied:
Also change function to a string.

modeswitch commented on test/core/configurator.js:
```diff
@@ -0,0 +1,602 @@
+
+    // Test load/store to/from local storage
+    test( 'local store/load', function() {
+        // Do not perform these tests if db access is not available
```
Don't disable these tests. We want to see when they fail!

modeswitch commented:
Make sure to test the case where DB access in unavailable to make sure your flag is set.

NorthWindH committed:
Merge remote-tracking branch 'ack/develop' into develop_issue_21

NorthWindH committed:
Updated nodb test message. Function is now a simple string.

NorthWindH committed:
Added test case that ensures that configurator's canUseDB is set correctly in a no db environment.
Minor moving around of db useability test code in configurator in order tobe more testable.
Small linting error addressed.

NorthWindH commented:
Changes implemented.
Review requested.

NorthWindH committed:
No longer throwing exceptions to error console.
Moved some code around.

NorthWindH committed:
Made test db name more random.

NorthWindH commented:
load/store and merge store tests may be passing when there is no db connectivity, this is bad behaviour

I could not reproduce this on my machine, @AlanKligman plz confirm? thanks! --@northWind87

NorthWindH committed:
Merge remote-tracking branch 'ack/develop' into develop_issue_21

Conflicts:
	src/gladius-src.js

NorthWindH committed:
Removed some spaces in gladius-src to match ack/develop

NorthWindH committed:
Clarified exception caught when window.indexedDB.open is tested

NorthWindH committed:
Minor linting

NorthWindH committed:
No db test now uses more random db name in initial assertion

NorthWindH committed:
Merge remote-tracking branch 'ack/develop' into develop_issue_21

NorthWindH committed:
Merge remote-tracking branch 'ack/develop' into develop_issue_21

NorthWindH committed:
Renamed incoming construction options.
Added incoming logger option.

NorthWindH committed:
Renamed logger to debug

NorthWindH committed:
Configurator now picks up default.js

NorthWindH committed:
Configurator now picks up default.js internally

NorthWindH committed:
Renamed parameters to options

NorthWindH committed:
Moved game id key into a const, now stored on configurator instance.
Game id key is now /id

NorthWindH committed:
Having weird problems. Code not working after make. Made investigation branch...

NorthWindH committed:
Turns out there were no real problems, my firefox profile had somehow become corrupted... In addition the indexeddb spec has changed, major changes will have to come in.
Removed some redundant debug messages.

NorthWindH committed:
Removed game id from default.js.
Added test game id to start up of engine if in debug. This can perhaps be improved to only account for the testing case as opposed to lumping debugging with testing.

NorthWindH committed:
Added error callbacks to load/store

NorthWindH committed:
storeJSON and getStoredJSON now both accept options object

NorthWindH committed:
_injectObjectStore now accepts options object

NorthWindH committed:
_injectDB and _ensureObjectStore now both accept options object

NorthWindH committed:
Added error handling to indexeddb hierarchy.

NorthWindH committed:
lint

modeswitch commented:
Tests are not running after merging your latest changes. Can you confirm that they run and pass for you?

NorthWindH committed:
Fixed oversight, error function is now passed into _getStoredJSON from load/store

NorthWindH commented:
@alankligman Hmmm I did some tests, the results are interesting.

### FF7
- **northwind/develop_issue_21:** works!
- **ack/develop:** works!

### FF8
- **northwind/develop_issue_21:** Broken. Configurator attempts to load, fails to acquire db, then calls gladius-src/initialize. Eventually threading/proxy throws an error to console ( "Could not get domain!" ) from threading/proxy.js Line 29
- **ack/develop:** Broken. No error message, did not investigate.

### Nightly ( firefox 11.0a1 (2011-12-09) )
- **northwind/develop_issue_21:** Broken. Configurator fails to initialize db ( "setVersion is not a function" ). Needs try/catch statement.
- **ack/develop:** Works!

Which version should we be aiming to please? :S

NorthWindH committed:
Threw try catches around all sensitive areas.
Added error handlers to all load/store calls in tests, now test suite proceeds even if configurator calls unsupported functionality.

NorthWindH committed:
minor

NorthWindH committed:
error was being used out of scope, corrected

NorthWindH commented:
**northwind/develop_issue_21** now works in nightly! setVersion is now nolonger a function so the load/store tests fail appropriately while the rest of the suite runs with no errors.

modeswitch commented:
If it works in FF9 that's good. It will need to work in nightly as well at some point, but that can be a different task.

NorthWindH commented:
I've tested the configurator branch in FF9; the test suite is not broken but the configurator load/store tests obviously don't pass because setVersion is no longer a function. I'll look into how the interface has changed and modify appropriately.

Also, the core/thread: Basic test never completes on the configurator branch on FF9. I've tested the ack/develop branch on FF9 and it did not have this issue. I traced the test code around a little on the configurator branch but didn't find any obvious problems. The last thing that happens is that the function ready() on pool.js gets called and goes down the path **( _queuedRequests.length > 0 ) = false => _readyThreads.push( thread )**

NorthWindH committed:
Pulled out getErrorFunc, now used ubiquitously

NorthWindH committed:
Slightly changed options acceptance style in functions that accept options object

NorthWindH committed:
Changed db version number, moved req into try

NorthWindH commented:
test suite should now run on FF9. Configurator tests should pass. @alankligman please confirm that tests run on your end?

You may have to clear the web storage for the tests page then reload the page. To do that, click the page's icon in the address bar, click more information, navigate to the permissions tab then click 'clear storage' under 'maintain offline storage'.

**Note:** core/thread Basic seems to pass after about... 5 minutes of waiting lol

modeswitch commented on src/core/configurator.js:
```diff
@@ -0,0 +1,525 @@
+            objectStoreName         = '',
+            gameId                  = '',
+
+            _getErrorFunc = function( error, funcName, msg ) {
```
This function doesn't seem to do any useful work. Why not just concatenate the strings and pass them into the error handler?

modeswitch commented on src/core/configurator.js:
```diff
@@ -0,0 +1,525 @@
+                        }
+                    }
+                } catch ( e ) {
+                    _getErrorFunc( error, '_ensureObjectStore', 'encountered exception' )( e );
```
This try block is way too large. Put it just around the code that might throw an exception and make the error message specific. "Encountered exception" is not acceptable.

modeswitch commented on src/core/configurator.js:
```diff
@@ -0,0 +1,525 @@
+
+        // Remove listener currently associated with client.
+        this.ignore = function() {
+            var curListener = this.node.listeners[this.id];
```
This variable assignment is unnecesary.

modeswitch commented on src/core/configurator.js:
```diff
@@ -0,0 +1,525 @@
+            });
+        };
+
+        defReadOnly( 'KEY_GAME_ID', KEY_GAME_ID );
```
This is also unnecessary. Just define the properties.

modeswitch commented on src/core/configurator.js:
```diff
@@ -0,0 +1,525 @@
+                };
+            },
+
+            _injectDB = function( options ) {
```
Rename this to openDB, or something a little more clear.

modeswitch commented on src/gladius-src.js:
```diff
@@ -25,7 +27,10 @@ define( function ( require ) {
+    global = window.gladius || ( window.gladius = {} ),
+
+    // Get testing game id ready
+    testGameID = window.guid();
```
This does not belong here. Remove it.

modeswitch commented on src/gladius-src.js:
```diff
@@ -34,120 +39,145 @@ define( function ( require ) {
+        // Load persistent configuration then proceed with rest of engine startup
+        this.configurator.load( {
+            callback: initialize,
+            error: initialize
```
This needs to be fixed. Initialize won't handle errors that come from the configurator.

NorthWindH replied:
At the moment, should load be unable to read from the db ie if db access is not available, the error handler is called. Since in this case the engine should load regardless, what would be the desired course of action?

NorthWindH committed:
Removed comment from configurator.js:L48

NorthWindH committed:
Removed _getErrorFunc

NorthWindH committed:
Broke down large try catch statements

NorthWindH committed:
Reformatted a try/catch

NorthWindH committed:
Removed redundant variable assignment

NorthWindH committed:
Flattened read only property access

NorthWindH committed:
Renamed _injectDB to _openDB

NorthWindH committed:
Gladius now accepts options.id to specify gameid.
Test gameid now created in test setup, nolonger inside gladius.

NorthWindH committed:
onerror handler for configurator on gladius construction

NorthWindH committed:
Updated configurator fail to load and fail to store messages

NorthWindH commented:
@alankligman 
Ready for review, everything fixed up.

NorthWindH committed:
Merge remote-tracking branch 'ack/develop' into develop_issue_21

Conflicts:
	src/gladius-src.js
	test/index.html
	test/index.html.dist
